Add environment variable query

madsmtm opened the pull request:
Generally, `rustc` prefers command-line arguments, but in some cases, an environment variable really is the most sensible option. We should make sure that this works properly with the compiler's change-tracking mechanisms, such that changing the relevant environment variable causes a rebuild.

This PR is a first step forwards in doing that.

Part of the work needed to do https://github.com/rust-lang/rust/issues/118204, see https://github.com/rust-lang/rust/pull/129342 for some discussion.

r? @petrochenkov

madsmtm commented on compiler/rustc_interface/src/passes.rs:
```diff
@@ -336,6 +337,20 @@ fn early_lint_checks(tcx: TyCtxt<'_>, (): ()) {
     )
 }
 
+fn env_var(tcx: TyCtxt<'_>, key: Symbol) -> Option<Symbol> {
```
I was unsure where to put the query definition, so I just put it in the same file as some other general queries.

madsmtm commented on compiler/rustc_interface/src/passes.rs:
```diff
@@ -336,6 +337,20 @@ fn early_lint_checks(tcx: TyCtxt<'_>, (): ()) {
 }
 
+fn env_var(tcx: TyCtxt<'_>, key: Symbol) -> Option<Symbol> {
+    let var = match std::env::var(key.as_str()) {
```
Is there something similar to Clippy's `disallowed-methods` that I could use to lint against usage of `std::env::var` in the compiler, and direct users towards the query instead?

See also https://github.com/rust-lang/cargo/issues/11588.

clubby789 replied:
There's some rustc-internal lints in https://github.com/rust-lang/rust/blob/master/compiler/rustc_lint/src/internal.rs

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11055112143/job/30713552240) [(plain)](https://github.com/rust-lang/rust/commit/506b0dfb219a0022d4a977f3097164bf67657e88/checks/30713552240/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 3.051 Building wheels for collected packages: reuse
#13 3.052   Building wheel for reuse (pyproject.toml): started
#13 3.317   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.318   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.318   Stored in directory: /tmp/pip-ephem-wheel-cache-phpx29yn/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.321 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.749 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.750 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.348 Collecting virtualenv
#13 4.348 Collecting virtualenv
#13 4.387   Downloading virtualenv-20.26.5-py3-none-any.whl (6.0 MB)
#13 4.469      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 75.7 MB/s eta 0:00:00
#13 4.517 Collecting distlib<1,>=0.3.7
#13 4.527   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.535      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 82.0 MB/s eta 0:00:00
#13 4.576 Collecting platformdirs<5,>=3.9.1
#13 4.580   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.624 Collecting filelock<4,>=3.12.2
#13 4.629   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.713 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.910 Successfully installed distlib-0.3.8 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.26.5
#13 DONE 5.0s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      192448 kB
DirectMap2M:     8196096 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
downloading https://static.rust-lang.org/dist/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
---
Diff in /checkout/compiler/rustc_interface/src/errors.rs:1:
+use std::ffi::OsString;
+use std::io;
 use std::path::Path;
-use std::{ffi::OsString, io};
 use rustc_macros::Diagnostic;
 use rustc_span::{Span, Symbol};
Diff in /checkout/compiler/rustc_borrowck/src/nll.rs:21:
 use rustc_mir_dataflow::move_paths::MoveData;
 use rustc_mir_dataflow::move_paths::MoveData;
 use rustc_mir_dataflow::points::DenseLocationMap;
 use rustc_session::config::MirIncludeSpans;
-use rustc_span::symbol::sym;
 use rustc_span::Symbol;
+use rustc_span::symbol::sym;
 use tracing::{debug, instrument};
 
 use crate::borrow_set::BorrowSet;
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/core/src/primitive.rs" "/checkout/compiler/rustc_borrowck/src/universal_regions.rs" "/checkout/compiler/rustc_borrowck/src/consumers.rs" "/checkout/compiler/rustc_borrowck/src/borrow_set.rs" "/checkout/compiler/rustc_borrowck/src/used_muts.rs" "/checkout/library/core/src/sync/atomic.rs" "/checkout/library/core/src/sync/mod.rs" "/checkout/library/core/src/sync/exclusive.rs" "/checkout/library/core/src/internal_macros.rs" "/checkout/library/core/src/str/converts.rs" "/checkout/library/core/src/str/error.rs" "/checkout/library/core/src/str/count.rs" "/checkout/library/core/src/str/iter.rs" "/checkout/library/core/src/str/lossy.rs" "/checkout/library/core/src/str/pattern.rs" "/checkout/library/core/src/str/validations.rs" "/checkout/library/core/src/str/mod.rs" "/checkout/library/core/src/str/traits.rs" "/checkout/compiler/rustc_borrowck/src/diagnostics/move_errors.rs" "/checkout/compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs" "/checkout/compiler/rustc_borrowck/src/diagnostics/region_errors.rs" "/checkout/compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs" "/checkout/compiler/rustc_borrowck/src/diagnostics/bound_region_errors.rs" "/checkout/compiler/rustc_borrowck/src/diagnostics/var_name.rs" "/checkout/compiler/rustc_borrowck/src/diagnostics/outlives_suggestion.rs" "/checkout/compiler/rustc_borrowck/src/diagnostics/find_use.rs" "/checkout/compiler/rustc_borrowck/src/diagnostics/mod.rs" "/checkout/compiler/rustc_borrowck/src/diagnostics/region_name.rs" "/checkout/compiler/rustc_borrowck/src/diagnostics/explain_borrow.rs" "/checkout/compiler/rustc_borrowck/src/diagnostics/find_all_local_uses.rs" "/checkout/compiler/rustc_borrowck/src/session_diagnostics.rs" "/checkout/compiler/rustc_borrowck/src/place_ext.rs" "/checkout/library/core/src/alloc/global.rs" "/checkout/library/core/src/alloc/layout.rs" "/checkout/library/core/src/alloc/mod.rs" "/checkout/library/core/src/default.rs" "/checkout/compiler/rustc_borrowck/src/region_infer/opaque_types.rs" "/checkout/compiler/rustc_borrowck/src/region_infer/dump_mir.rs" "/checkout/compiler/rustc_borrowck/src/region_infer/reverse_sccs.rs" "/checkout/library/core/src/net/display_buffer.rs" "/checkout/compiler/rustc_borrowck/src/region_infer/values.rs" "/checkout/library/core/src/net/parser.rs" "/checkout/compiler/rustc_borrowck/src/region_infer/mod.rs" "/checkout/library/core/src/net/ip_addr.rs" "/checkout/compiler/rustc_borrowck/src/region_infer/graphviz.rs" "/checkout/library/core/src/net/mod.rs" "/checkout/library/core/src/net/socket_addr.rs" "/checkout/compiler/rustc_borrowck/src/dataflow.rs" "/checkout/compiler/rustc_borrowck/src/def_use.rs" "/checkout/library/core/src/pat.rs" "/checkout/compiler/rustc_borrowck/src/location.rs" "/checkout/compiler/rustc_borrowck/src/lib.rs" "/checkout/compiler/rustc_borrowck/src/member_constraints.rs" "/checkout/compiler/rustc_borrowck/src/borrowck_errors.rs" "/checkout/compiler/rustc_borrowck/src/nll.rs" "/checkout/library/core/src/unicode/unicode_data.rs" "/checkout/library/core/src/unicode/printable.rs" "/checkout/library/core/src/unicode/mod.rs" "/checkout/library/core/src/ub_checks.rs" "/checkout/library/core/src/asserting.rs" "/checkout/library/core/src/hint.rs" "/checkout/compiler/rustc_borrowck/src/type_check/relate_tys.rs" "/checkout/compiler/rustc_borrowck/src/type_check/input_output.rs" "/checkout/compiler/rustc_borrowck/src/places_conflict.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Thu Sep 26 15:16:28 UTC 2024
  network time: Thu, 26 Sep 2024 15:16:28 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11059152683/job/30726825626) [(plain)](https://github.com/rust-lang/rust/commit/af26a40e14cbbf7eb982ebdc0cb8cfcdf5ac549d/checks/30726825626/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.801 Building wheels for collected packages: reuse
#13 2.802   Building wheel for reuse (pyproject.toml): started
#13 3.052   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.053   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.054   Stored in directory: /tmp/pip-ephem-wheel-cache-g1tcv0pz/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.057 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.458 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.458 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.024 Collecting virtualenv
#13 4.024 Collecting virtualenv
#13 4.080   Downloading virtualenv-20.26.5-py3-none-any.whl (6.0 MB)
#13 4.282      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 30.2 MB/s eta 0:00:00
#13 4.343 Collecting platformdirs<5,>=3.9.1
#13 4.352   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.376 Collecting distlib<1,>=0.3.7
#13 4.397      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 58.3 MB/s eta 0:00:00
#13 4.438 Collecting filelock<4,>=3.12.2
#13 4.446   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.446   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.527 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.718 Successfully installed distlib-0.3.8 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.26.5
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      190400 kB
DirectMap2M:     8198144 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://static.rust-lang.org/dist/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
---
##[endgroup]
fmt check
fmt: checked 5594 files
tidy check
tidy error: /checkout/compiler/rustc_interface/messages.ftl: message `interface_temps_dir_error` appears before `interface_env_var_not_unicode`, but is alphabetically later than it
run `./x.py test tidy --bless` to sort the file correctly
removing old virtual environment
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.2)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Thu Sep 26 19:50:27 UTC 2024
  network time: Thu, 26 Sep 2024 19:50:27 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11072138881/job/30765697778) [(plain)](https://github.com/rust-lang/rust/commit/0ef957071dbccbbad7701f06e3e34b3dc2b598e8/checks/30765697778/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.891 Building wheels for collected packages: reuse
#16 2.892   Building wheel for reuse (pyproject.toml): started
#16 3.142   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.143   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.144   Stored in directory: /tmp/pip-ephem-wheel-cache-11eq74zm/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.146 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.542 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.542 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.6s
---
    Checking rustc_borrowck v0.0.0 (/checkout/compiler/rustc_borrowck)
error: unused import: `env`
 --> compiler/rustc_borrowck/src/nll.rs:6:11
  |
6 | use std::{env, io};
  |
  = note: `-D unused-imports` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(unused_imports)]`

```

</details>

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11073815647/job/30771214236) [(plain)](https://github.com/rust-lang/rust/commit/455e468a8072254bce86ab350a10f8a386e8332f/checks/30771214236/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.816 Building wheels for collected packages: reuse
#16 2.817   Building wheel for reuse (pyproject.toml): started
#16 3.074   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.074   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.075   Stored in directory: /tmp/pip-ephem-wheel-cache-3rxoik9x/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.077 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.477 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.477 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.6s
---
    Checking rustc_interface v0.0.0 (/checkout/compiler/rustc_interface)
error: referenced message `key` does not exist (in message `interface_env_var_not_unicode`)
  --> compiler/rustc_interface/src/lib.rs:25:40
   |
25 | rustc_fluent_macro::fluent_messages! { "../messages.ftl" }
   |
   |
   = help: you may have meant to use a variable reference (`{$key}`)

error: referenced message `var` does not exist (in message `interface_env_var_not_unicode`)
   |
   |
25 | rustc_fluent_macro::fluent_messages! { "../messages.ftl" }
   |
   |
   = help: you may have meant to use a variable reference (`{$var}`)
error: could not compile `rustc_interface` (lib) due to 2 previous errors
warning: build failed, waiting for other jobs to finish...
Build completed unsuccessfully in 0:02:26
  local time: Fri Sep 27 16:06:58 UTC 2024
```

</details>

madsmtm commented on compiler/rustc_interface/src/passes.rs:
```diff
@@ -336,6 +337,20 @@ fn early_lint_checks(tcx: TyCtxt<'_>, (): ()) {
+            None
+        }
+    };
+    // Also add the variable to Cargo's dependency tracking
+    tcx.sess.psess.env_depinfo.borrow_mut().insert((key, var));
```
Hmm, a complication here is that this query will only work with the dependency info files that Cargo reads if it's called _before_ `passes::write_dep_info` at: https://github.com/rust-lang/rust/blob/a3f76a26e045a760bb1163b7eab36872985242d5/compiler/rustc_driver_impl/src/lib.rs#L437

So everything done during `passes::analysis` and `Linker::codegen_and_build_linker` will not be able to use this to, and will need some other mechanism for populating `ParseSess::env_depinfo`.

Is it possible for us to move the `write_dep_info` pass to the very end, or just before linking? Or is it important that it happens as soon as possible for parallelizing Cargo, or something?

petrochenkov replied:
>Or is it important that it happens as soon as possible for parallelizing Cargo, or something?

It is important for depinfo to be generated as early as possible, without performing full compilation.
All *file* dependencies in particular, and env var dependencies from `env!` and proc macros are ready after name resolution is completed.

I think it's fine to still lose some env depinfo in this PR, it will still better than what we have now.
Could you add a comment to this code, explaining that env vars added after the call to `write_dep_info` will be lost from the dep file?

(We can think what to do with the Apple env vars specifically later.)

madsmtm replied:
I have added a comment to the `env_var` query, and added `rustc_codegen_ssa::register_environment_variables` as an extension point for where to put this. Suggestions very welcome on how to do it differently!

petrochenkov commented on compiler/rustc_borrowck/src/nll.rs:
```diff
@@ -21,6 +21,7 @@ use rustc_mir_dataflow::impls::MaybeInitializedPlaces;
 use rustc_mir_dataflow::points::DenseLocationMap;
 use rustc_session::config::MirIncludeSpans;
+use rustc_span::Symbol;
 use rustc_span::symbol::sym;
```
```suggestion
use rustc_span::{sym, Symbol};
```

petrochenkov commented on compiler/rustc_middle/src/query/mod.rs:
```diff
@@ -118,6 +118,16 @@ rustc_queries! {
+    /// detection and target-dependent compiler flags.
+    ///
+    /// Will emit an error and return `None` if the variable is not UTF-8.
+    query env_var(key: Symbol) -> Option<Symbol> {
```
Is it possible to use `OsStr(ing)`s in query inputs and outputs? What are the obstacles?
Many places in the compiler use `env::var_os`.

Also, `Symbol`s shouldn't be needed here, you can take a `OsStr`/`str` and return `&'tcx OsString`/`&'tcx String` like `env::var(_os)` returns.

madsmtm replied:
`OsStr` isn't really set up with the query system, so we'd have to do that first. I've done parts of it in the last commit, but it also needs to be `Key`, which I'm unsure would be correct? `&str` for example isn't `Key` either.

Also, returning `&'tcx OsStr` requires that the environment variables are actually stored inside `TyCtxt<'tcx>`, which I'm unsure how to do wrt. lifetimes?

petrochenkov replied:
>Also, returning `&'tcx OsStr` requires that the environment variables are actually stored inside `TyCtxt<'tcx>`, which I'm unsure how to do wrt. lifetimes?

Things can be stored in inside `TyCtxt` using the `tcx.arena.alloc*` methods.
A lot of queries return references to something allocated on tcx arenas.

petrochenkov replied:
>I've done parts of it in the last commit, but it also needs to be `Key`, which I'm unsure would be correct?

I don't know what exactly needs to be implemented to support OS strings in queries.
I'll need to try doing that myself to answer questions like this, for which I probably don't have time at the moment.

madsmtm replied:
I got it to work, thanks for the tips!

petrochenkov commented on compiler/rustc_middle/src/query/mod.rs:
```diff
@@ -118,6 +118,16 @@ rustc_queries! {
         desc { "perform lints prior to macro expansion" }
     }
 
+    /// Useful for the implementation of `std::env!`, `proc-macro`s change
```
It should mention "tracked access to environment variables" as the first thing in the comment.

bors commented:
:umbrella: The latest upstream changes (presumably #132581) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

madsmtm committed:
Test that env! works with incremental compilation

This currently works because it's part of expansion, and that isn't yet
tracked by the query system. But we want to ensure it continues working,
even if that is changed.

madsmtm commented:
@rustbot ready

petrochenkov commented on compiler/rustc_interface/src/passes.rs:
```diff
@@ -647,6 +666,10 @@ pub fn write_dep_info(tcx: TyCtxt<'_>) {
+    // Similarly, allow codegen and linking passes to state which environment
+    // variables they depend on, as those passes are run after this pass, but
+    // we'll need the information when emitting dependency info to Cargo.
+    rustc_codegen_ssa::register_environment_variables(tcx);
```
I don't think it makes sense to add this until we are actually doing something there.

madsmtm commented on compiler/rustc_middle/src/query/mod.rs:
```diff
@@ -121,6 +122,21 @@ rustc_queries! {
+    /// NOTE: This currently does not work with dependency info in the
+    /// analysis, codegen and linking passes, place extra code at the top of
+    /// `rustc_interface::passes::write_dep_info` to make that work.
+    query env_var(key: &'tcx OsStr) -> Option<&'tcx OsStr> {
```
Now that this returns `OsStr`, should it be called `env_var_os`? And then we can add a helper function on the side that converts into a string if UTF-8?

petrochenkov replied:
I agree.

madsmtm replied:
Have created one now in `rustc_middle/src/util/mod.rs`, though unsure if there's a better place for it.

madsmtm reviewed (commented):
@rustbot ready

Note that I decided to remove the "Add environment variable tracking in places where it was convenient" commit, since they weren't working properly anyway without the extra support code in `emit_dep_info` (and then I'd rather have people add that support code when they need it).

petrochenkov commented on compiler/rustc_middle/src/query/mod.rs:
```diff
@@ -121,6 +122,21 @@ rustc_queries! {
+    /// NOTE: This currently does not work with dependency info in the
+    /// analysis, codegen and linking passes, place extra code at the top of
+    /// `rustc_interface::passes::write_dep_info` to make that work.
+    query env_var(key: &'tcx OsStr) -> Option<&'tcx OsStr> {
```
```suggestion
    query env_var(key: &OsStr) -> Option<&'tcx OsStr> {
```
Does the key have to be `'tcx`?
The implementation logic doesn't require it.

If the query infra requires it, then it's better to make a wrapper that allocates it on `tcx` arena, otherwise it will be inconvenient to use it.

madsmtm replied:
> Does the key have to be `'tcx`?

Yup.

> it's better to make a wrapper that allocates it on `tcx` arena, otherwise it will be inconvenient to use it.

In most cases, the key will be `'static` though.

petrochenkov commented on compiler/rustc_data_structures/src/stable_hasher.rs:
```diff
@@ -565,6 +565,12 @@ where
+impl_stable_traits_for_trivial_type!(::std::ffi::OsString);
+
+impl_stable_traits_for_trivial_type!(::std::ffi::CStr);
+impl_stable_traits_for_trivial_type!(::std::ffi::CString);
```
Is it required to add the impls for `CStr(ing)` and `OsString` for `OsStr` to work?

Same question about the stuff in `compiler/rustc_middle/src/query/keys.rs`.

madsmtm replied:
Nope, I just thought I'd add them while I was at it. I can remove them if you think they're not desirable

petrochenkov commented on compiler/rustc_interface/src/passes.rs:
```diff
@@ -321,6 +321,31 @@ fn early_lint_checks(tcx: TyCtxt<'_>, (): ()) {
 }
 
+fn env_var<'tcx>(tcx: TyCtxt<'tcx>, key: &'tcx OsStr) -> Option<&'tcx OsStr> {
+    let value = std::env::var_os(key);
```
```suggestion
    let value = env::var_os(key);
```
Style nit: could you import instead of using 3+ segment paths?

petrochenkov commented on compiler/rustc_interface/src/passes.rs:
```diff
@@ -321,6 +321,31 @@ fn early_lint_checks(tcx: TyCtxt<'_>, (): ()) {
+    tcx.sess.psess.env_depinfo.borrow_mut().insert((
+        Symbol::intern(&key.to_string_lossy()),
+        value.map(|value| Symbol::intern(&value.to_string_lossy())),
+    ));
```
Maybe skip this step if the string is not UTF-8, instead of putting the incorrect values into the map?

In a similar situation `tracked_env::var` in `library\proc_macro\src\lib.rs` resets non-UTF-8 *values* into `None`, perhaps unintentionally.

Not sure what would be the best recovery strategy here.

petrochenkov replied:
It would probably be better to turn `psess.env_depinfo` into `FxIndexSet<(OsString, Option<OsString>)>` and delay the recovery until `fn write_out_deps`.
Maybe use `to_string_lossy` there, just make sure the lossy values don't accidentally collide with the non-lossy ones.

Not in this PR though.

madsmtm replied:
I've changed this code to do the same as in `tracked_env::var` for now, and have opened https://github.com/rust-lang/rust/issues/133699 to track using `OsStr` in `env_depinfo`

petrochenkov commented:
> @rustbot ready
> 
> Note that I decided to remove the "Add environment variable tracking in places where it was convenient" commit, since they weren't working properly anyway without the extra support code in `emit_dep_info` (and then I'd rather have people add that support code when they need it).

But they did work, for incremental scenarios? Just not for cargo rebuilds.
I'd rather brought those changes back.

madsmtm committed:
Add environment variable tracking in places where it was convenient

This won't work with Cargo's change tracking, but it should work with incremental.

madsmtm commented:
@rustbot ready

> > Note that I decided to remove the "Add environment variable tracking in places where it was convenient" commit, since they weren't working properly anyway without the extra support code in `emit_dep_info` (and then I'd rather have people add that support code when they need it).
> 
> But they did work, for incremental scenarios? Just not for cargo rebuilds. I'd rather brought those changes back.

I think so. I've brought them back.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12104859025/job/33748766705) [(plain)](https://github.com/rust-lang/rust/commit/6ca63b32679318b9c8fdcd452425a90c2518a1b1/checks/33748766705/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 27.3s done
#16 DONE 32.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---

error[E0424]: expected value, found module `self`
  --> compiler/rustc_middle/src/util/mod.rs:38:30
   |
37 |     pub fn env_var(key: impl AsRef<OsStr>) -> Result<&'tcx str, VarError> {
   |            ------- this function doesn't have a `self` parameter
38 |         if let Some(value) = self.env_var_os(key.as_ref()) {
   |                              ^^^^ `self` value is a keyword only available in methods with a `self` parameter
   |
help: add a `self` receiver parameter to make the associated `fn` a method
   |
37 |     pub fn env_var(&self, key: impl AsRef<OsStr>) -> Result<&'tcx str, VarError> {

   Compiling rustc_ast_passes v0.0.0 (/checkout/compiler/rustc_ast_passes)
   Compiling rustc_expand v0.0.0 (/checkout/compiler/rustc_expand)
   Compiling rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12105111748/job/33749301095) [(plain)](https://github.com/rust-lang/rust/commit/745b34e150aaf3e156327915d2176e28d86b4124/checks/33749301095/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 27.1s done
#16 DONE 30.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
   Compiling rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error[E0599]: no method named `into_string` found for reference `&std::ffi::OsStr` in the current scope
  --> compiler/rustc_middle/src/util/mod.rs:41:19
   |
41 |             value.into_string().map_err(VarError::NotUnicode)
   |
help: there is a method `to_str` with a similar name
   |
   |
41 |             value.to_str().map_err(VarError::NotUnicode)

For more information about this error, try `rustc --explain E0599`.
error: could not compile `rustc_middle` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12105860020/job/33750960599) [(plain)](https://github.com/rust-lang/rust/commit/1028b1e6e84d92a220f9c2fe22896c13f64e1e8a/checks/33750960599/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#17 exporting to docker image format
#17 sending tarball 27.4s done
#17 DONE 33.3s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
   Compiling rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error[E0599]: no method named `map_err` found for enum `std::option::Option` in the current scope
    --> compiler/rustc_middle/src/util/mod.rs:41:28
     |
41   |             value.to_str().map_err(VarError::NotUnicode)
     |
help: there is a method `map_or` with a similar name, but with different arguments
    --> /checkout/library/core/src/option.rs:1167:5
     |
     |
1167 | /     pub fn map_or<U, F>(self, default: U, f: F) -> U
1169 | |         F: FnOnce(T) -> U,
     | |__________________________^

For more information about this error, try `rustc --explain E0599`.
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12105923144/job/33751101403) [(plain)](https://github.com/rust-lang/rust/commit/aa575adcea574a40083650ffe085aaab0783f0f3/checks/33751101403/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.0s done
#16 DONE 34.6s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
   Compiling rustc_builtin_macros v0.0.0 (/checkout/compiler/rustc_builtin_macros)
error[E0597]: `key` does not live long enough
  --> compiler/rustc_middle/src/util/mod.rs:40:46
   |
38 | impl<'tcx> TyCtxt<'tcx> {
   |      ---- lifetime `'tcx` defined here
39 |     pub fn env_var(&self, key: impl AsRef<OsStr>) -> Result<&'tcx str, VarError> {
   |                           --- binding `key` declared here
40 |         if let Some(value) = self.env_var_os(key.as_ref()) {
   |                              ----------------^^^----------
   |                              |               borrowed value does not live long enough
   |                              argument requires that `key` is borrowed for `'tcx`
...
45 |     }
45 |     }
   |     - `key` dropped here while still borrowed

error[E0309]: the parameter type `impl AsRef<OsStr>` may not live long enough
   |
38 | impl<'tcx> TyCtxt<'tcx> {
38 | impl<'tcx> TyCtxt<'tcx> {
   |      ---- the parameter type `impl AsRef<OsStr>` must be valid for the lifetime `'tcx` as defined here...
39 |     pub fn env_var(&self, key: impl AsRef<OsStr>) -> Result<&'tcx str, VarError> {
40 |         if let Some(value) = self.env_var_os(key.as_ref()) {
   |                                              ^^^^^^^^^^^^ ...so that the type `impl AsRef<OsStr>` will meet its required lifetime bounds
help: consider adding an explicit lifetime bound
   |
   |
39 |     pub fn env_var(&self, key: impl AsRef<OsStr> + 'tcx) -> Result<&'tcx str, VarError> {

Some errors have detailed explanations: E0309, E0597.
For more information about an error, try `rustc --explain E0309`.
error: could not compile `rustc_middle` (lib) due to 2 previous errors
```

</details>

bjorn3 commented on compiler/rustc_borrowck/src/nll.rs:
```diff
@@ -145,9 +145,11 @@ pub(crate) fn compute_regions<'a, 'tcx>(
+            let algorithm = infcx
+                .tcx
+                .env_var_os("POLONIUS_ALGORITHM".as_ref())
+                .unwrap_or_else(|| Symbol::intern("Hybrid"));
```
I think a better solution here would be to add a `polonius_algorithm` field to `Options` and read the `POLONIUS_ALGORITHM` env var in `build_session_options`. Then marking the field as `[TRACKED]` is enough to get it tracked by incr comp.

madsmtm replied:
That is _exactly_ what I did originally in https://github.com/rust-lang/rust/pull/129342, but @petrochenkov [preferred](https://github.com/rust-lang/rust/pull/129342#issuecomment-2335153699) that we integrated it with the query system instead, hence this PR.

I have no idea what's the best solution here, I'll defer to you two to discuss and decide on that ;) .

bjorn3 replied:
I did prefer accessing env vars as early as possible to prevent eg proc-macros from affecting compilation options. Proc-macros are known to do things they should not be allowed to do.

petrochenkov replied:
Having a query for env vars probably still makes sense even if some of the env var uses are converted to options.

(One downside of early tracking is that the dependency on the env var will be registered even if the env var is not actually used.)

petrochenkov commented on compiler/rustc_middle/src/util/mod.rs:
```diff
@@ -29,3 +34,13 @@ impl std::ops::Deref for Providers {
     }
 }
+
+impl<'tcx> TyCtxt<'tcx> {
```
Maybe put it to `compiler\rustc_middle\src\ty\context.rs`?
That's where `TyCtxt` is defined and many of its random inherent methods live.

petrochenkov commented on compiler/rustc_middle/src/util/mod.rs:
```diff
@@ -29,3 +34,13 @@ impl std::ops::Deref for Providers {
 }
+
+impl<'tcx> TyCtxt<'tcx> {
+    pub fn env_var(self, key: &'tcx OsStr) -> Result<&'tcx str, VarError> {
```
Can use `AsRef<OsStr>` here, then the function will accept string literals.
You even use `key.as_ref()` below, which is unnecessary if the key is a non-generic `OsStr`.

petrochenkov commented on compiler/rustc_middle/src/util/mod.rs:
```diff
@@ -29,3 +34,13 @@ impl std::ops::Deref for Providers {
+
+impl<'tcx> TyCtxt<'tcx> {
+    pub fn env_var(self, key: &'tcx OsStr) -> Result<&'tcx str, VarError> {
+        if let Some(value) = self.env_var_os(key.as_ref()) {
```
Style nit: if-lets with short (one line) arms typically look better as a match.

petrochenkov commented on compiler/rustc_middle/src/query/keys.rs:
```diff
@@ -492,6 +494,22 @@ impl Key for Option<Symbol> {
+    }
+}
+
+impl<'tcx> Key for Option<&'tcx OsStr> {
```
Is this necessary?
`Option<&'tcx OsStr>` doesn't seem to be used as a key.

petrochenkov commented on compiler/rustc_middle/src/query/erase.rs:
```diff
@@ -255,6 +264,7 @@ trivial! {
     Option<rustc_span::def_id::DefId>,
     Option<rustc_span::def_id::LocalDefId>,
     Option<rustc_span::Span>,
+    Option<rustc_span::Symbol>,
```
This is likely not necessary anymore.

petrochenkov commented on compiler/rustc_middle/src/query/erase.rs:
```diff
@@ -180,6 +185,10 @@ impl<T> EraseType for Option<&'_ [T]> {
     type Result = [u8; size_of::<Option<&'static [()]>>()];
 }
 
+impl EraseType for Option<&'_ OsStr> {
```
Probably unnecessary.

madsmtm replied:
It's necessary because of the return type in `TyCtxt::env_var_os`.

bors commented:
:umbrella: The latest upstream changes (presumably #135438) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

petrochenkov commented:
Still waiting on the author to address the review.
@madsmtm Do you plan to return to this?

madsmtm commented:
Yeah, I think I got a bit discouraged by https://github.com/rust-lang/rust/pull/130883#discussion_r1864919351, so I feared that we wouldn't be able to land it because of disagreements. But I seem to have misremembered, so thanks for the ping, I'll try to pick it up one of the coming days.

madsmtm commented:
Finally got around to rebasing and addressing your review comments.

@rustbot ready

petrochenkov commented on compiler/rustc_borrowck/src/nll.rs:
```diff
@@ -163,8 +163,8 @@ pub(crate) fn compute_regions<'a, 'tcx>(
             let algorithm =
-                env::var("POLONIUS_ALGORITHM").unwrap_or_else(|_| String::from("Hybrid"));
-            let algorithm = Algorithm::from_str(&algorithm).unwrap();
+                infcx.tcx.env_var("POLONIUS_ALGORITHM").unwrap_or_else(|_| "Hybrid".as_ref());
```
```suggestion
                infcx.tcx.env_var("POLONIUS_ALGORITHM").unwrap_or_else(|_| "Hybrid");
```

petrochenkov commented on compiler/rustc_interface/src/passes.rs:
```diff
@@ -672,6 +697,9 @@ pub fn write_dep_info(tcx: TyCtxt<'_>) {
     let _ = tcx.resolver_for_lowering();
+    // Similarly, analysis, codegen and linking should state which environment
+    // variables they depend on here, as those passes are run after this pass,
+    // but we'll need the information when emitting dependency info to Cargo.
```
Could you remove this comment until it's actually done?
I'm not sure it will be done in the way that is described here.

madsmtm committed:
Add `TyCtx::env_var_os`

Along with `TyCtx::env_var` helper. These can be used to track
environment variable accesses in the query system.

Since `TyCtx::env_var_os` uses `OsStr`, this commit also adds the
necessary trait implementations for that to work.

madsmtm commented:
Done!

@rustbot ready

petrochenkov commented:
Thanks!
@bors r+

bors commented:
:pushpin: Commit 632ce38c9a93a6ee890aedd99fcb1a61cabd0a46 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 632ce38c9a93a6ee890aedd99fcb1a61cabd0a46 -->
<!-- homu: {"type":"Approved","sha":"632ce38c9a93a6ee890aedd99fcb1a61cabd0a46","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 632ce38c9a93a6ee890aedd99fcb1a61cabd0a46 with merge ecb170afc878648c3ae355dbd596c8e4b6f7ebdc...
<!-- homu: {"type":"BuildStarted","head_sha":"632ce38c9a93a6ee890aedd99fcb1a61cabd0a46","merge_sha":"ecb170afc878648c3ae355dbd596c8e4b6f7ebdc"} -->
